third_party/amd/lib/TritonAMDGPUTransforms/StreamPipeline.cpp

```diff
@@ -197,53 +209,69 @@ class StreamPipeliner {
 //   WARNING: Changing the order of schedule.clusters.newAtBack() calls
 //            can cause invalid schedules to be produced.
 void StreamPipeliner::initSchedule(int maxIndirectionLevel) {
-  int lastStage = numStages - 1;
-  config.resize(SCHED_TAIL + 1);
-
-  bool isMultibuf = numStages > (2 + maxIndirectionLevel);
-  if (prefetch) {
-    // Prefetch Schema cluster order and staging.
-    // for i in (...):
-    //   local_stores: stage=i+1
-    //   global_loads: stage=i+2
-    //   compute:      stage=i
-    //   local_load:   stage=i+1
-    //   tail:         stage=i
-    config[SCHED_LOCAL_STORE] = {lastStage - 1, schedule.clusters.newAtBack()};
-    auto cluster1 = schedule.clusters.newAtBack();
-    config[SCHED_GLOBAL_LOAD] = {0, cluster1};
-    config[SCHED_COMPUTE] = {lastStage, cluster1};
-    config[SCHED_LOCAL_LOAD] = {lastStage - 1, schedule.clusters.newAtBack()};
-    config[SCHED_TAIL] = {lastStage, schedule.clusters.newAtBack()};
-  } else if (isMultibuf) {
-    // Streaming Schema cluster order and staging for multi-buffer.
-    // for i in (...):
-    //   local_stores: stage=i+1
-    //   global_loads: stage=i+2
-    //   local_load:   stage=i
-    //   compute:      stage=i
-    //   tail:         stage=i
-    config[SCHED_LOCAL_STORE] = {lastStage - 1, schedule.clusters.newAtBack()};
-    auto cluster1 = schedule.clusters.newAtBack();
-    config[SCHED_GLOBAL_LOAD] = {0, cluster1};
-    config[SCHED_LOCAL_LOAD] = {lastStage, cluster1};
-    config[SCHED_COMPUTE] = {lastStage, cluster1};
-    config[SCHED_TAIL] = {lastStage, schedule.clusters.newAtBack()};
+
+  bool pairedGlobalLoadLocalStore = stages[SCHED_LOCAL_STORE] == 0;
+  if (stages[SCHED_LOCAL_STORE] < 0) {
+    stages[SCHED_LOCAL_STORE] = numStages + stages[SCHED_LOCAL_STORE];
   } else {
-    // Streaming Schema cluster order and staging for single-buffer.
-    // for i in (...):
-    //   global_loads: stage=i+1
-    //   local_load:   stage=i
-    //   compute:      stage=i
-    //   local_stores: stage=i+1
-    //   tail:         stage=i
-    auto cluster0 = schedule.clusters.newAtBack();
-    config[SCHED_GLOBAL_LOAD] = {0, cluster0};
-    config[SCHED_LOCAL_LOAD] = {lastStage, schedule.clusters.newAtBack()};
-    config[SCHED_COMPUTE] = {lastStage, cluster0};
-    config[SCHED_LOCAL_STORE] = {lastStage - 1, schedule.clusters.newAtBack()};
-    config[SCHED_TAIL] = {lastStage, schedule.clusters.newAtBack()};
+    stages[SCHED_LOCAL_STORE] += maxIndirectionLevel;
+  }
+
+  if (stages[SCHED_LOCAL_LOAD] < 0)
+    stages[SCHED_LOCAL_LOAD] = numStages + stages[SCHED_LOCAL_LOAD];
+
+  assert(stages[SCHED_GLOBAL_LOAD] <= stages[SCHED_LOCAL_STORE] <=
+         stages[SCHED_LOCAL_LOAD]);
```

Comment:
This needs to be separated as a <=b && b<=c. As an example, it produces false assert error when all of three values are 0.